File: src/Cryptography/WringTwistree/Permute.hs
module Cryptography.WringTwistree.Permute
  ( permut8
  , permute256
  ) where

{- This module is used in both Wring and Twistree.
 - It is part of the keying algorithm, which turns a byte string
 - into three s-boxes. permSbox takes a sequence of 96 Word16 (the high
 - bit is ignored) and returns a permutation of [0x00..0xff], which is
 - one of the three s-boxes.
 -}

import Data.Bits
import Data.Array.Unboxed
import Data.Word
import qualified Data.Sequence as Seq
import Data.Sequence ((><), (<|), (|>), Seq((:<|)), Seq((:|>)), update)

swapOrder :: Word16 -> [Int]
swapOrder n = map fromIntegral [x2,x3,x4,x5,x6,x7,x8] where
  x2 = n .&. 1
  x4 = (n `div` 2) .&. 3
  x8 = (n `div` 8) .&. 7
  x18 = (n `div` 64) .&. 15 + 1
  x3 = x18 `mod` 3
  x6 = x18 `div` 3
  x35' = (n `div` 1024) .&. 31 + 1
  x35 = if x35' > 16 then x35' + 1 else x35'
  x5 = x35 `mod` 5
  x7 = x35 `div` 5

swapmute :: Seq.Seq a -> [Int] -> Int -> Seq.Seq a
swapmute ys [] _ = ys
swapmute ys (x:xs) n = swapmute ys' xs (n+1) where
  b = Seq.index ys x
  c = Seq.index ys n
  ys' = update x c (update n b ys)

permut8 :: Seq.Seq a -> Word16 -> Seq.Seq a
permut8 ys n = swapmute ys (swapOrder n) 1

permut8x32 :: Seq.Seq Word16 -> Seq.Seq a -> Seq.Seq a
permut8x32 _ Seq.Empty = Seq.Empty
permut8x32 Seq.Empty sbox = sbox -- this should never happen
permut8x32 key sbox = permHead >< permTail where
  permHead = permut8 (Seq.take 8 sbox) (Seq.index key 0)
  permTail = permut8x32 (Seq.drop 1 key) (Seq.drop 8 sbox)

-- 10 is a primitive root of 257; 180 is its inverse. The eight bytes in a group
-- all go to and come from different groups of eight.
dealInxA :: Int -> Int
dealInxA n = ((n+1)*10) `mod` 257-1

invDealInxA :: Int -> Int
invDealInxA n = ((n+1)*180) `mod` 257-1

-- polynomial 100011101, 3 bit overflow table
shift3    = listArray (0,7) [0x00,0x1d,0x3a,0x27,0x74,0x69,0x4e,0x53] :: UArray Int Int
invShift3 = listArray (0,7) [0x00,0xad,0x47,0xea,0x8e,0x23,0xc9,0x64] :: UArray Int Int

dealInxB :: Int -> Int
dealInxB n = ((n .&. 0x1f) `shiftL` 3) `xor` (shift3 ! ((n .&. 0xe0) `shiftR` 5))

invDealInxB :: Int -> Int
invDealInxB n = ((n .&. 0xf8) `shiftR` 3) `xor` (invShift3 ! (n .&. 0x07))

dealInx = dealInxB
invDealInx = invDealInxB

dealBytes :: Seq.Seq a -> Seq.Seq a -- must be 256 long
dealBytes bs = Seq.fromList $ map (Seq.index bs) $ map invDealInx [0..255]

permute256 :: Seq.Seq Word16 -> Seq.Seq Word8
permute256 k = dealBytes $ permut8x32 k2 $ dealBytes $ permut8x32 k1 $
               dealBytes $ permut8x32 k0 $ Seq.fromList [0..255] where
  k012 = Seq.chunksOf 32 k
  k0 = Seq.index k012 0
  k1 = Seq.index k012 1
  k2 = Seq.index k012 2

File: src/permute.rs
// Module permute
// This module is used in both Wring and Twistree.
// It permutes an array of 256 bytes by an array of 96 15-bit numbers,
// resulting in an s-box.

fn permut8(ys: &mut [u8], n: u16) {
  let mut swap_order:[u16; 7]=[0;7];
  swap_order[0]=n&1;
  swap_order[2]=(n>>1)&3;
  swap_order[6]=(n>>3)&7;
  let mut temp=((n>>6)&15)+1;
  swap_order[1]=temp%3;
  swap_order[4]=temp/3;
  temp=((n>>10)&31)+1;
  if temp>16 {temp+=1;}
  swap_order[3]=temp%5;
  swap_order[5]=temp/5;
  for i in 0..7 {
    ys.swap(i+1,swap_order[i] as usize);
  }
}

#[test]
fn test_permut8() {
  let mut word:[u8; 8]=*b"calipers";
  permut8(&mut word[..],0x3c45);
  assert_eq!(&word,b"spiracle");
  word=*b"recounts";
  permut8(&mut word[..],0x595b);
  assert_eq!(&word,b"construe");
  word=*b"thousand";
  permut8(&mut word[..],0x30da);
  assert_eq!(&word,b"handouts");
  for i in 0..0x8000 {
    word=*b"repaints";
    permut8(&mut word[..],i);
    assert_ne!(&word,b"pantries");
  }
}

fn permut8x32(sbox: &mut [u8], key: &[u16]) {
  assert_eq!(sbox.len(),8*key.len());
  for i in 0..key.len() {
    permut8(&mut sbox[8*i..8*i+8],key[i]);
  }
}

const SHIFT3: [u16;8]=[0x00,0x1d,0x3a,0x27,0x74,0x69,0x4e,0x53];

fn deal_inx(n:usize) -> usize {
  ((n<<3)&0xf8)^(SHIFT3[(n>>5)&7]) as usize
}

fn deal_bytes(sbox0: &[u8],sbox1: &mut [u8]) {
  assert_eq!(sbox0.len(),256);
  assert_eq!(sbox1.len(),256);
  for i in 0..256 {
    sbox1[deal_inx(i)]=sbox0[i];
  }
}

pub fn permute256(sbox: &mut [u8],key: &[u16]) {
  assert_eq!(sbox.len(),256);
  assert_eq!(key.len(),96);
  for i in 0..256 { sbox[i]=i as u8; }
  let mut sbox2:[u8;256]=[0;256];
  let mut sbox3:[u8;256]=[0;256];
  permut8x32(sbox,&key[0..32]);
  deal_bytes(&sbox,&mut sbox2);
  permut8x32(&mut sbox2,&key[32..64]);
  deal_bytes(&sbox2,&mut sbox3);
  permut8x32(&mut sbox3,&key[64..96]);
  deal_bytes(&sbox3,sbox);
}
